core/jobs/io/gcs_io.py

```diff
@@ -0,0 +1,269 @@
+
+MYPY = False
+if MYPY:  # pragma: no cover
+    from apache_beam.io.gcp import gcsio_test
+    from mypy_imports import app_identity_services
+
+app_identity_services = models.Registry.import_app_identity_services()
+
+
+# TODO(#15613): Here we use MyPy ignore because of the incomplete typing of
+# apache_beam library and absences of stubs in Typeshed, forces MyPy to
+# assume that PTransform class is of type Any. Thus to avoid MyPy's error (Class
+# cannot subclass 'PTransform' (has type 'Any')), we added an ignore here.
+class ReadFile(beam.PTransform): # type: ignore[misc]
+    """Read files form the GCS."""
+
+    def __init__(
+        self,
+        client: Optional[gcsio_test.FakeGcsClient] = None,
+        mode: str = 'r',
+        label: Optional[str] = None
+    ) -> None:
+        """Initializes the ReadFile PTransform.
+
+        Args:
+            client: Optional[gcsio_test.FakeGcsClient]. The GCS client to use
+                while testing. This will be None when testing on server.
+            mode: str. The mode in which the file will be read, can be
+                'r' and 'rb'.
+            label: Optional[str]. The label of the PTransform.
+        """
+        super().__init__(label=label)
+        self.client = client
+        self.mode = mode
+
+    def expand(self, file_paths: beam.PCollection) -> beam.PCollection:
+        """Returns PCollection with file data.
+
+        Args:
+            file_paths: PCollection. The collection of filepaths that will
+                be read.
+
+        Returns:
+            PCollection. The PCollection of the file data.
+        """
+        return (
+            file_paths
+            | 'Read the file' >> beam.Map(self._read_file)
+        )
+
+    def _read_file(self, file_path: str) -> Union[bytes, str]:
+        """Helper function to read the contents of a file.
+
+        Args:
+            file_path: str. The name of the file that will be read.
+
+        Returns:
+            data: Union[bytes, str]. The file data.
+        """
+        gcs = gcsio.GcsIO(self.client)
+        bucket = app_identity_services.get_gcs_resource_bucket_name()
+        gcs_url = f'gs://{bucket}/{file_path}'
+        file = gcs.open(gcs_url, mode=self.mode)
+        data = file.read()
+        file.close()
+        return data
+
+
+class FileObjectDict(TypedDict):
+    """Dictionary representing file object that will be written to GCS."""
+
+    filepath: str
+    data: Union[bytes, str]
+
+
+# TODO(#15613): Here we use MyPy ignore because of the incomplete typing of
+# apache_beam library and absences of stubs in Typeshed, forces MyPy to
+# assume that PTransform class is of type Any. Thus to avoid MyPy's error (Class
+# cannot subclass 'PTransform' (has type 'Any')), we added an ignore here.
+class WriteFile(beam.PTransform): # type: ignore[misc]
+    """Write files to GCS."""
+
+    def __init__(
+        self,
+        client: Optional[gcsio_test.FakeGcsClient] = None,
```

Comment:
I think we should have a comment explaining this.

Reply:
(why we pass fake client in non-test file)

Reply:
Done